src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1777 @@
+    omap_inner_key_t node_key;
+    node_key.laddr = laddr;
+    node_key.key_len = val.size() + 1;
+    node_key.key_off = iter.get_index() == 0 ?
+                       node_key.key_len :
+                       (iter - 1).get_node_key().key_off + node_key.key_len;
+    if (recorder) {
+      recorder->insert(
+        node_key,
+        val);
+    }
+    inner_insert(iter, node_key, val);
+  }
+
+  void journal_inner_update(
+    const_iterator _iter,
+    const laddr_t laddr,
+    delta_inner_buffer_t *recorder) {
+    auto iter = iterator(this, _iter.index);
+    auto node_key = iter.get_node_key();
+    node_key.laddr = laddr;
+    if (recorder) {
+      recorder->update(node_key, iter->get_node_val());
+    }
+    inner_update(iter, node_key);
+  }
+
+  void journal_inner_replace(
+    const_iterator _iter,
+    const laddr_t laddr,
+    const std::string val,
+    delta_inner_buffer_t *recorder) {
+    auto iter = iterator(this, _iter.index);
+    omap_inner_key_t node_key;
+    node_key.laddr = laddr;
+    node_key.key_len = val.size() + 1;
+    node_key.key_off = iter.get_index() == 0?
+                       node_key.key_len :
+                       (iter - 1).get_node_key().key_off + node_key.key_len;
+    if (recorder) {
+      recorder->remove(iter->get_node_val());
+      recorder->insert(node_key, val);
+    }
+    inner_replace(iter, node_key, val);
+  }
+
+  void journal_inner_remove(
+    const_iterator _iter,
+    delta_inner_buffer_t *recorder) {
+    auto iter = iterator(this, _iter.index);
+    if (recorder) {
+      recorder->remove(iter->get_node_val());
+    }
+    inner_remove(iter);
+  }
+
+  StringKVInnerNodeLayout(char *buf) :
+    buf(buf) {}
+
+  uint32_t get_size() const {
+    ceph_le32 &size = *layout.template Pointer<0>(buf);
+    return uint32_t(size);
+  }
+
+  /**
+   * set_size
+   *
+   * Set size representation to match size
+   */
+  void set_size(uint32_t size) {
+    ceph_le32 s;
+    s = size;
+    *layout.template Pointer<0>(buf) = s;
+  }
+
+  const_iterator iter_begin() const {
+    return const_iterator(
+      this,
+      0);
+  }
+
+  const_iterator iter_end() const {
+    return const_iterator(
+      this,
+      get_size());
+  }
+
+  iterator iter_begin() {
+    return iterator(
+      this,
+      0);
+  }
+
+  iterator iter_end() {
+    return iterator(
+      this,
+      get_size());
+  }
+
+  const_iterator iter_idx(uint16_t off) const {
+    return const_iterator(
+      this,
+      off);
+  }
+
+  const_iterator string_lower_bound(std::string str) const {
+    uint16_t start = 0, end = get_size();
+    while (start != end) {
+      unsigned mid = (start + end) / 2;
+      const_iterator iter(this, mid);
+      std::string s = iter->get_node_val();
+      if (s < str)
+        start = ++mid;
+      if ( s > str)
+        end = mid;
+      if (s == str)
+        return iter;
+    }
+    return const_iterator(this, start);
+  }
+
+  iterator string_lower_bound(std::string str) {
```

Comment:
const ref -- actually, check every string param in this file and convert to const ref if appopriate.  We do want to avoid copying strings unless necessary.

Reply:
string_view everywhere?